Documentation/specs/corehost.md

```diff
@@ -0,0 +1,112 @@
+# corehost runtime/assembly resolution
+
+The shared host locates assemblies and native libraries using a combination of: Servicing Index, Files in the application folder (aka "app-local") and files from package caches.
+
+## Definitions and Formats
+
+### Servicing Index
+
+The servicing index is loaded when the `DOTNET_SERVICING` environment variable is non-empty. When this variable is non-empty, it points to a directory that will be the **Servicing Root**.
+
+An index file is located at the path defined by `$DOTNET_SERVICING/servicing_index.txt`. In this file are a series of lines of one of the following formats:
+
+```
+# Lines starting with a '#' and blank lines are ignored.
+
+# Identifies an asset from a NuGet Package that has been serviced.
+package|[Package ID]|[Package Version]|[Original Asset Relative Path]=[New Asset Path, relative to Servicing Root]
+package|System.Threading.Thread|1.2.3.4|lib/dotnet5.4/System.Threading.Thread.dll=patches/abc123/System.Threading.Thread.dll
+
+# TBD: Host/Runtime servicing entries.
+```
+
+Paths in this file are **always** specified using `/`, even on Windows. They must be converted to platform-specific directory separators.
+
+This index is loaded when needed during the Resolution Process (see below).
+
+### Dependencies File
```

Comment:
A `.deps` file is not mandatory, but it is required for servicing. We could make it mandatory, but I think that if a user really wants to put everything app-local and remove the `.deps` file, they should be able to. Right now, `publish` does in fact remove the `.deps` file, since servicing wasn't fully worked out and thus it didn't need the file, but we will update it to keep the `.deps` file around.

A long-term goal would be to attempt to embed the deps file into the host (there are a bunch of issues there such as varying binary formats and signing) which would make the deps file invisible to the user.